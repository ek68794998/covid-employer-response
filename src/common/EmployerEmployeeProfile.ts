import { EmployeeNumber } from "./EmployeeNumber";

export type EmployerEmployeeProfileType =
	"range"
	| "approximately"
	| "exactly";

export class EmployerEmployeeProfile {
	public lowerBound?: number;

	public type: EmployerEmployeeProfileType = "exactly";

	public upperBound: number = 0;

	public year: number = 0;

	public yearQuarter?: "Q1" | "Q2" | "Q3" | "Q4";

	public static toString(p: EmployerEmployeeProfile, includeDate: boolean, useShortText: boolean): string | null {
		const lowerBound: EmployeeNumber | null =
			(p.lowerBound && p.lowerBound > 0) ? new EmployeeNumber(p.lowerBound) : null;

		const upperBound: EmployeeNumber | null =
			p.upperBound > 0 ? new EmployeeNumber(p.upperBound) : null;

		let employeeCountString: string;

		switch (p.type) {
			case "approximately":
				if (!upperBound) {
					throw new Error("Invalid value for approximate string conversion.");
				}

				employeeCountString = upperBound.toString(useShortText);

				if (!upperBound.getMagnitudeLabel()) {
					employeeCountString = `~${employeeCountString}`;
				}

				break;

			case "range":
				if (lowerBound && upperBound) {
					if (lowerBound.exactValue > upperBound.exactValue) {
						throw new Error("Lower bound is greater than or equal to upper bound specified for range string conversion.");
					}

					employeeCountString =
						(lowerBound.magnitude && lowerBound.magnitude === upperBound.magnitude)
							? `${lowerBound.getAdjustedValue()} ${String.fromCharCode(0x2013)} ${upperBound.toString(useShortText)}`
							: `${lowerBound.toString(useShortText)} ${String.fromCharCode(0x2013)} ${upperBound.toString(useShortText)}`;
				} else if (lowerBound) {
					employeeCountString =
						useShortText
							? `${String.fromCharCode(0x2265)} ${lowerBound.toString(useShortText)}`
							: `More than ${lowerBound.toString(useShortText)}`;
				} else if (upperBound) {
					employeeCountString =
						useShortText
							? `${String.fromCharCode(0x2264)} ${upperBound.toString(useShortText)}`
							: `Less than ${upperBound.toString(useShortText)}`;
				} else {
					throw new Error("Invalid lower and upper bounds specified for range string.");
				}

				break;

			case "exactly":
			default:
				if (!upperBound) {
					throw new Error("Invalid value for default string conversion.");
				}

				employeeCountString = upperBound.toString(useShortText);

				break;
		}

		if (includeDate && p.year > 0) {
			const dateString: string =
				p.yearQuarter ? `${p.yearQuarter} ${p.year}` : `${p.year}`;

			employeeCountString = `${employeeCountString} (${dateString})`;
		}

		return employeeCountString;
	}
}
