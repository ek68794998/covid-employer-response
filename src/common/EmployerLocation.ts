export class EmployerLocation {
	public city: string = "";

	public country: string = "";

	public international: boolean = false;

	public state?: string;

	public wiki?: string;

	public static toString(
		employerLocation: EmployerLocation,
		countryCodeMap?: { [key: string]: string }): string {

		const country: string =
			(countryCodeMap && employerLocation.country in countryCodeMap)
				? countryCodeMap[employerLocation.country]
				: employerLocation.country;

		const locationAsString: string =
			employerLocation.state
				? `${employerLocation.city}, ${employerLocation.state}, ${country}`
				: `${employerLocation.city}, ${country}`;

		return employerLocation.international ? `${locationAsString} (Multinational)` : locationAsString;
	}
}