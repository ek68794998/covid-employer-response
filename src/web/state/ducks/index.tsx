import { combineReducers } from "redux";

import { getLocalizedStrings } from "./localization/reducers";

export default combineReducers({
	strings: getLocalizedStrings,
});