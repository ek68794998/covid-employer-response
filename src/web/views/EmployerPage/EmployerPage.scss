@import "../Colors.scss";
@import "../Dimensions.scss";

main#employer-page {
	margin: 0 auto;
	padding: 0 $page-padding;
	width: 100%;

	> div {
		margin: 0 auto;
	}
	
	div.EmployerPage__Content {
		max-width: $body-max-width;
		padding-bottom: 16px;
		padding-top: 16px;
	}
}

@media (max-width: $body-narrow-width) {
	main#employer-page {
		padding: 0 $page-padding-narrow;
	}
}