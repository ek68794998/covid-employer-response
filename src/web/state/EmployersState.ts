import { EmployerRecord } from "../../common/EmployerRecord";

export interface EmployersState {
	items: EmployerRecord[];
}