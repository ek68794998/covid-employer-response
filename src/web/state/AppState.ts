import { EmployersState } from "./EmployersState";

import { LocalizedStrings } from "../../common/LocalizedStrings";

export interface AppState {
	employers: EmployersState;

	environment: string;

	strings: LocalizedStrings;
}