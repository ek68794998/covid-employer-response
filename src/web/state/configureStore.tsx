import { AnyAction, applyMiddleware, compose, createStore, Store } from "redux";
import thunk from "redux-thunk";

import { Production } from "../../common/constants/EnvironmentConstants";

import { AppState } from "./AppState";
import rootReducer from "./ducks";

const configureStore = (preloadedState: Partial<AppState>): Store<AppState, AnyAction> => {
	const isProd: boolean = process.env.NODE_ENV === Production;
	const composeEnhancers: any =
		(!isProd && typeof window !== "undefined" && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

	const store: Store<AppState, AnyAction> = createStore(
		rootReducer,
		preloadedState,
		composeEnhancers(applyMiddleware(thunk)),
	);

	if (module.hot) {
		// Enable Webpack hot module replacement for reducers
		module.hot.accept("./ducks", () => {
			const nextRootReducer: any = require("./ducks").default;
			store.replaceReducer(nextRootReducer);
		});
	}

	return store;
};

export default configureStore;