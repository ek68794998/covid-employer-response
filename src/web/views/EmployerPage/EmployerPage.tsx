import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { RouteComponentProps } from "react-router-dom";

import { EmployerRecord } from "../../../common/EmployerRecord";

import { AppState } from "../../state/AppState";
import { getEmployer } from "../../state/ducks/employers/selectors";

import EmployerPageDetails from "../EmployerPageDetails/EmployerPageDetails";

import "./EmployerPage.scss";
import { getEmployerById } from "../../state/ducks/employers/actions";

interface Params {
	id: string;
}

type Props = RouteComponentProps<Params>;

const EmployerPage: React.FC<Props> = (props: Props): React.ReactElement => {
	const dispatch: React.Dispatch<any> = useDispatch();

	const employerId: string = props.match.params.id;

	const employer: EmployerRecord | undefined =
		useSelector((state: AppState) => getEmployer(state, employerId));

	useEffect(
		() => {
			dispatch(getEmployerById(employerId));
		},
		[ employerId ]);

	if (!employer) {
		return (
			<main id="employer-page">
				<div className="EmployerPage__Content">
					Not found.
				</div>
			</main>
		);
	}

	return (
		<main id="employer-page">
			<div className="EmployerPage__Content">
				<EmployerPageDetails employer={employer} />
			</div>
		</main>
	);
};

export default EmployerPage;
