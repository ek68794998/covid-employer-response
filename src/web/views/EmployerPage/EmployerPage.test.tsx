import React from "react";
import * as ReactRedux from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";
import { AnyAction, Store } from "redux";

import { getPlocStringsAsync, mockComponent } from "../../../__tests__/TestUtils";

import { EmployerEmployeeProfile } from "../../../common/EmployerEmployeeProfile";
import { EmployerLocation } from "../../../common/EmployerLocation";
import { EmployerRecord } from "../../../common/EmployerRecord";
import { EmployerRecordBase } from "../../../common/EmployerRecordBase";
import { EmployerRecordMetadata } from "../../../common/EmployerRecordMetadata";
import { AppState } from "../../state/AppState";
import configureStore from "../../state/configureStore";

import EmployerPage from "./EmployerPage";

jest.mock(
	"../EmployerPageDetails/EmployerPageDetails",
	() => mockComponent("EmployerPageDetails"));

jest.mock(
	"../LoadingIndicator/LoadingIndicator",
	() => mockComponent("LoadingIndicator"));

describe("<EmployerPage />", () => {
	beforeEach(() => {
		jest.spyOn(ReactRedux, "useDispatch").mockReturnValue(jest.fn());
	});

	afterEach(() => {
		(ReactRedux.useDispatch as any).mockRestore();
	});

	test("renders without exploding", async () => {
		const store: Store<AppState, AnyAction> = configureStore({ strings: await getPlocStringsAsync() });

		const renderedValue: ReactTestRendererJSON | null =
			renderer.create(
				<ReactRedux.Provider store={store}>
					<BrowserRouter>
						<EmployerPage employerId="foo" />
					</BrowserRouter>
				</ReactRedux.Provider>,
			).toJSON();

		expect(renderedValue).toMatchSnapshot();
	});

	test("displays an employer record", async () => {
		const fakeEmployer: Partial<EmployerRecordBase> = {
			employeesBefore: new EmployerEmployeeProfile(),
			id: "e1",
			location: new EmployerLocation(),
			name: "Alpha",
		};

		const store: Store<AppState, AnyAction> = configureStore({
			employers: {
				itemsComplete: {
					[fakeEmployer.id as string]: Object.assign(new EmployerRecord(), fakeEmployer),
				},
				itemsMetadata: {
					[fakeEmployer.id as string]: Object.assign(new EmployerRecordMetadata(0, 0, 0, "fair"), fakeEmployer),
				},
			},
			strings: await getPlocStringsAsync(),
		});

		const renderedValue: ReactTestRendererJSON | null =
			renderer.create(
				<ReactRedux.Provider store={store}>
					<BrowserRouter>
						<EmployerPage employerId={fakeEmployer.id || ""} />
					</BrowserRouter>
				</ReactRedux.Provider>,
			).toJSON();

		expect(renderedValue).toMatchSnapshot();
	});
});
