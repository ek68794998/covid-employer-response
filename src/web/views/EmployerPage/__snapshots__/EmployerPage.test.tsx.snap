// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmployerPage /> displays an employer record 1`] = `
<main
  id="employer-page"
>
  <div
    className="EmployerPage__Content"
  >
    [component: EmployerPageDetails employer=([object Object]) /]
  </div>
</main>
`;

exports[`<EmployerPage /> renders without exploding 1`] = `
<main
  id="employer-page"
>
  <div
    className="EmployerPage__Content"
  >
    Not found.
  </div>
</main>
`;
